rslib/i18n_helpers/src/garbage_collection.rs

```diff
@@ -0,0 +1,222 @@
+            && entry
+                .file_name()
+                .to_str()
+                .expect("non-unicode filename")
+                .ends_with(".ftl")
+        {
+            let ftl = fs::read_to_string(entry.path()).expect("failed to open file");
+            let mut ast = parser::parse(ftl.as_str()).expect("failed to parse ftl");
+            extract_nested_ftl_messages(&mut &ast, &mut used_ftls);
+            let num_entries = ast.body.len();
+            ast.body = ast
+                .body
+                .into_iter()
+                .filter(|entry| match entry {
+                    ast::Entry::Message(msg) => used_ftls.contains(msg.id.name),
+                    _ => true,
+                })
+                .collect();
+            if ast.body.len() < num_entries {
+                fs::write(entry.path(), serialize::serialize(&ast)).expect("failed to write file");
+            }
+        }
+    }
+}
+
+fn extract_nested_ftl_messages(ast: &ast::Resource<&str>, messages: &mut HashSet<String>) {
+    for entry in &ast.body {
+        if let ast::Entry::Message(ast::Message {
+            value: Some(pattern),
+            ..
+        }) = entry
+        {
+            extract_messages_from_pattern(pattern, messages);
+        }
+    }
+}
+
+fn extract_messages_from_pattern(pattern: &ast::Pattern<&str>, messages: &mut HashSet<String>) {
+    for elem in &pattern.elements {
+        if let ast::PatternElement::Placeable { expression } = elem {
+            extract_messages_from_expression(expression, messages);
+        }
+    }
+}
+
+fn extract_messages_from_expression(
+    expression: &ast::Expression<&str>,
+    messages: &mut HashSet<String>,
+) {
+    match expression {
+        ast::Expression::Select { selector, variants } => {
+            extract_messages_from_inline_expression(selector, messages);
+            for variant in variants {
+                extract_messages_from_pattern(&variant.value, messages);
+            }
+        }
+        ast::Expression::Inline(expression) => {
+            extract_messages_from_inline_expression(expression, messages)
+        }
+    }
+}
+
+fn extract_messages_from_inline_expression(
+    expression: &ast::InlineExpression<&str>,
+    messages: &mut HashSet<String>,
+) {
+    match expression {
+        ast::InlineExpression::MessageReference { id, .. } => {
+            messages.insert(id.name.to_string());
+        }
+        ast::InlineExpression::Placeable { expression } => {
+            extract_messages_from_expression(expression, messages)
+        }
+        _ => (),
+    }
+}
+
+fn extract_references_from_file(refs: &mut HashSet<String>, entry: &DirEntry) {
+    lazy_static! {
+        static ref RUST_STYLE_TR: Regex = Regex::new(r"\Wtr\s*\.([0-9a-z_]+)\W").unwrap();
+        static ref TS_STYLE_TR: Regex = Regex::new(r"\Wtr2?\.([0-9A-Za-z_]+)\W").unwrap();
+        static ref DESIGNER_STYLE_TR: Regex = Regex::new(r"<string>([0-9a-z_]+)</string>").unwrap();
```

Comment:
Some things that could be improved here. I don't think `\W` is the best choice to find a name delimiter in any of these languages and maybe at least the last two regexes should be verbose.

Reply:
I'm not too fussed about these - verbose would be a bit more readable, but as they're fairly short, it's not super-important. \W seems to be working - can you come up with an example of a case we're likely to hit in the future where it would not be appropriate?

Reply:
I was thinking more about the opposite, because \W is a unicode class and will match all sorts of stuff that we don't need. But I'm probably just being pedantic. \W is much better to understand and maintain than any explicit set of characters we could come up with, so never mind. 🙂